Статистика выполнения кода

dmpas opened the pull request:
* Добавлен ключ `-codestat=<filename>`
 * Подсчёт строк приведён в соответствие с поведением Родительской Платформы. В качестве побочного явления - обилие LineNum в байткоде.
 * Подсчёт количества проходов и времени выполнения.

dmpas committed:
CodeStat: помечен заголовок метода.

Строка, на которой начинается объявление процедуры, обозначается в
байт-коде.

dmpas committed:
CodeStat: добавлена обработка Если и Пока.

dmpas committed:
Appveyor: выкладка архива сборки

dmpas committed:
CodeStat: Исправлено определение номера строки.

Номер строки теперь определяется на шаге расчленения текста, а не на
шаге компиляции, что  предотвращает случаи, когда парсер ускакивает на
следующую строку.

dmpas committed:
CodeStat: Исправлен подсчёт КонецЕсли

dmpas committed:
CodeStat: Добавлен подсчёт циклов

dmpas committed:
CodeStat: Обработка концов методов.

ВАЖНО: изменение в байт-коде. Теперь оператор "Возврат" не выполняет
непосредственный возврат из метода, а переходит в конец метода на
единственную в методе команду Return.

dmpas committed:
CodeStat: обработка Попытки.

dmpas committed:
AppVeyor: увеличен номер выпуска

dmpas committed:
CodeStat: Добавлен вывод невызывавшихся строк.

dmpas committed:
CodeStat: Исправлено определение номеров строк.

Были ошибки при определении номера строки в случаях если файл начинался с перевода строки LF.

dmpas committed:
Исправление опечатки.

dmpas committed:
CodeStat: Добавлен замер времени выполнения.

dmpas committed:
CodeStat: Добавлен простой тест.

dmpas commented:
@nixel2007 с тебя тесты.

nixel2007 commented on tests/codestat.os:
```diff
@@ -0,0 +1,230 @@
+		Возврат Неопределено;
+	КонецЕсли;
+
+	Если Чтение.ТипТекущегоЗначения = ТипЗначенияJson.НачалоОбъекта Тогда
```
Я буду еретиком, если спрошу можно ли перевести с встроенного потокового чтения на библиотеку json, которая сразу объект вернет? :(

artbear replied:
У меня был этот же вопрос.
Почему-то @dmpas не юзает библиотеки :(

dmpas replied:
Просто не хочется завязывать тестирование движка на библиотеки. А так, религия не запрещает.

artbear commented on tests/codestat.os:
```diff
@@ -0,0 +1,230 @@
+
+Функция ПолучитьВыводДляСкрипта(Знач ТекстСкрипта)
+
+	ИмяФайлаОСкрипта = ПолучитьИмяВременногоФайла("os");
```
Временные файлы специально не удаляются?
Предлагаю добавить их удаление в ПослеЗапускаТеста или использовать пакет ВременныеФайлы

dmpas replied:
Принято.

artbear commented on tests/codestat.os:
```diff
@@ -0,0 +1,230 @@
+Функция ПрочитатьЗначение(Знач Чтение)
+
+	Если Не Чтение.Прочитать() Тогда
+		Возврат Неопределено;
```
Предлагаю сразу выбрасывать исключение

dmpas replied:
Принято.

dmpas committed:
CodeStat: поправлен тест.

dmpas committed:
CodeStat: Упрощено добавление тестов.

dmpas committed:
CodeStat: Исправлен формат вывода.

Свойство, содержащее путь к файлу теперь называется `#path`, чтобы
избежать неверного вывода в случае, если модуль содержит метод `path()`.

nixel2007 commented:
@dmpas правильно понимаю, что новый тест теперь - просто новый os файлик в папке?
👍

dmpas commented:
@nixel2007 да, только разметить его надо соответствующим образом.

nixel2007 commented:
while.os
```bsl
Пока Истина Цикл     //- $entry: 1
	А = 0;           //- $entry: 1
	Прервать;        //- $entry: 1
	Б = 0;           //- $entry: 0
КонецЦикла;          //- $entry: 1

Для сч = 1 По 2 Цикл //- $entry: 3
	А = 0;           //- $entry: 2
КонецЦикла;          //- $entry: 3
```

try.os
```bsl
Попытка             //- $entry: 1
	А = 0;          //- $entry: 1
Исключение          //- $entry: 0
	Б = 0;          //- $entry: 0
КонецПопытки;       //- $entry: 1

Попытка             //- $entry: 1
	А = Число("ы"); //- $entry: 1
Исключение          //- $entry: 1
	Б = 0;          //- $entry: 1
КонецПопытки;       //- $entry: 1
```

p.s. я хз как коммитить прямо к тебе, кроме как новым PR. но помимо этого я не смогу прогнать тесты, у меня не собирается односкрипт(

dmpas commented:
@nixel2007 ты можешь всё тот же артефакт с аппвеёра качать.

dmpas commented:
@nixel2007 после "Прервать" КонецЦикла не выполняется. Ни у нас, ни в Жёлтой Маме. Потому, там будет 0.

Аналогично с Для - КонецЦикла будет выполнен только 2 раза. Воспринимай команду "КонецЦикла" как переход в начало и проверку условия. Если цикл закончен или прерван - КонецЦикла не выполняется.

dmpas committed:
CodeStat: добавлены тесты.

dmpas committed:
CodeStat: Исправлен посчёт цикла Пока

EvilBeaver commented on src/ScriptEngine/Compiler/Parser.cs:
```diff
@@ -233,23 +235,26 @@ public override Lexem ReadNextLexem(ParseIterator iterator)
                             Token = LanguageDef.GetToken(content),
-                            Content = content
+                            Content = content,
+                            LineNumber = currentLine
```
ой-вей... а оно точно надо в лексему номер строки добавлять?

dmpas replied:
Да, так гораздо проще. Потому что парсер после лексемы мог убежать уже куда угодно и слишком много пришлось бы тогда пилить в парсере. А тут просто запомнил, где встретилась лексема, и всё.

EvilBeaver commented on src/ScriptEngine/Compiler/SourceCodeIndexer.cs:
```diff
@@ -14,9 +14,9 @@ namespace ScriptEngine.Compiler
     {
         private readonly string _code;
-        private readonly IList<int> _lineBounds;
+        private readonly IDictionary<int, int> _lineBounds;
```
А как будет работать данный словарь? Чем старый вариант не угодил?

dmpas replied:
Разница при MoveBack/MoveNext. В List делался Add, что вообще не гарантировало, что нужная строка окажется именно по тому индексу, который ей соответствует. Лучше всего это ловилось на случаях, когда файл начинается с пустой строки - она вообще выпадала из анализа.
А в словаре в этом отношении всё однозначно.

EvilBeaver commented on src/ScriptEngine/Compiler/Compiler.cs:
```diff
@@ -689,7 +730,8 @@ private void BuildForEachStatement()
             BuildExpression(Token.Loop);
             AddCommand(OperationCode.PushIterator, 0);
-            var loopBegin = AddCommand(OperationCode.IteratorNext, 0);
+            var loopBegin = AddCommand(OperationCode.LineNum, _lastExtractedLexem.LineNumber);
```
Я немного потерялся в вашей дискуссии с @nixel2007. Если я правильно понял, то есть цель учитывать выполнение строк КонецЦикла и т.п?

Изначально опкод LineNumber использовался только для выдачи корректных рантайм исключений. Я верно понимаю, что сейчас предлагается его использовать еще и для статистики кода?

dmpas replied:
Да, верно. Сейчас порядок выполнения опкодов LineNumber совпадает с тем, как в 1С отладчик по шагам проходит.

dmpas commented:
@nixel2007 напомню про тесты, а то у меня под новый год будет затык со временем.

nixel2007 commented:
@dmpas я если честно не знаю, что ещё можно добавить. Вызовы функций есть, циклы, попытки, если есть.

artbear commented:
Рекурсию считали?

artbear commented:
+ Пока цикл
+ Для по Цикл
+ Для каждого Цикл
+ Если Тогда Иначе
+ Оператор ? как будет учтен?
+ (Новый Объект).ВызовМетода ?
+ Процедура/функция?

dmpas committed:
CodeStat: Указана кодировка JSON

dmpas committed:
CodeStat: убрана накрутка внутри одной строки.

Если оп-код `LineNum` устанавливает тот же самый номер строки, то
увеличения счётчика проходов теперь не происходит.

dmpas committed:
CodeStat: добавлены тесты foreach и subs.

nixel2007 commented:
@dmpas не могу понять, как запустить отдельно тест codestat.os
пишу `testrunner tests\codestat.os`, кладу новый файлик в папку с тестами
new.os
```bsl
А = Новый ТаблицаЗначений();           //- $entry: 1
Б = Новый ("ТаблицаЗначений");         //- $entry: 1

В = (Новый ТаблицаЗначений()).Колонки //- $entry: 1
	.Количество(); //- $entry: 1
```

независимо от цифр тестраннер всегда выдает зеленую полосу. запуск `oscript tests\testrunner.os tests\codestat.os` так же не приносит желаемого эффекта.
Что делаю не так?

dmpas commented:
@nixel2007 
```
testrunner -run codestat.os
```
так попробуй

artbear commented:
>не могу понять, как запустить отдельно тест codestat.os
пишу testrunner tests\codestat.os, кладу новый файлик в папку с тестами

@nixel2007 А доку почитать? :) или хотя бы хелп из командной строки?

dmpas commented:
@nixel2007 
```bsl
В = (Новый ТаблицаЗначений()).Колонки //- $entry: 1
	.Количество(); //- $entry: 1
```
тут сейчас, к сожалению (или радости) будет "1" и "0". Многострочные выражения что у нас, что в Жёлтой Маме всегда относятся на первую строку.

nixel2007 commented:
> тут сейчас, к сожалению (или радости) будет "1" и "0". Многострочные выражения что у нас, что в Жёлтой Маме всегда относятся на первую строку.

В "Жёлтой Маме" ( ахахахахха ) есть хитрость.
```bsl
МояФункция(            //- $entry: 1
    Истина,                  //- $entry: 0
    Истина,                  //- $entry: 0
    ДругаяФункция(),  //- $entry: 1
    Истина                   //- $entry: 0
);                                 //- $entry: 0
```

nixel2007 commented:
> testrunner -run codestat.os
так попробуй

получилось без -run. Оказывается, был важен CWD - нужно быть в tests

dmpas commented:
@nixel2007 
```bsl
  ДругаяФункция(),  //- $entry: 1
```
вот эту хитрость я, пожалуй, отработаю всё-таки.

nixel2007 commented:
@dmpas обрати внимание

![image](https://cloud.githubusercontent.com/assets/1132840/21453432/0087a7b8-c920-11e6-87bf-54d023b6245b.png)

В первом запуске вместо МояФункция() была платформенная СтрШаблон() --- замер не отметился на этой строке. А вот на собственной функции - отметился

dmpas committed:
CodeStat: Подсчёт вызова функции.

Вызов функции в многострочном выражении подсчитывается независимо от
корневого выражения.

dmpas commented:
Снял WIP. 15-ю версию выпустили, 16-я далеко, можно ломать ночники :smiley:
